src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+ *      Handle to information about the operation to be appended.
+ */
+void
+ClientTransactionTask::DecisionRpc::appendOp(CommitCacheMap::iterator opEntry)
+{
+    const CacheKey* key = &opEntry->first;
+    CacheEntry* entry = &opEntry->second;
+
+    request.emplaceAppend<WireFormat::TxParticipant>(
+            key->tableId,
+            static_cast<uint64_t>(key->keyHash),
+            entry->rpcId);
+
+    entry->state = CacheEntry::DECIDE;
+    ops[reqHdr->participantCount] = opEntry;
+    reqHdr->participantCount++;
+}
+
+/**
+ * Handle the case where the RPC may have been sent to the wrong server.
+ */
+void
+ClientTransactionTask::DecisionRpc::retryRequest()
+{
+    for (uint32_t i = 0; i < reqHdr->participantCount; i++) {
+        const CacheKey* key = &ops[i]->first;
+        CacheEntry* entry = &ops[i]->second;
+        ramcloud->objectFinder.flush(key->tableId);
+        entry->state = CacheEntry::PENDING;
+    }
+    task->nextCacheEntry = task->commitCache.begin();
+}
+
+/**
+ * Constructor for PrepareRpc.
+ *
+ * \param ramcloud
+ *      The RAMCloud object that governs this RPC.
+ * \param session
+ *      Session on which this RPC will eventually be sent.
+ * \param task
+ *      Pointer to the transaction task that issued this request.
+ */
+ClientTransactionTask::PrepareRpc::PrepareRpc(RamCloud* ramcloud,
+        Transport::SessionRef session, ClientTransactionTask* task)
+    : RpcWrapper(sizeof(WireFormat::TxPrepare::Response))
+    , ramcloud(ramcloud)
+    , task(task)
+    , ops()
+    , reqHdr(allocHeader<WireFormat::TxPrepare>())
+{
+    reqHdr->lease = task->lease;
+    reqHdr->participantCount = task->participantCount;
+    reqHdr->opCount = 0;
+    request.appendExternal(&task->participantList);
+    this->session = session;
+}
+
+// See RpcWrapper for documentation.
+bool
+ClientTransactionTask::PrepareRpc::checkStatus()
+{
+    if (responseHeader->status == STATUS_UNKNOWN_TABLET) {
+        retryRequest();
+    }
+    return true;
+}
+
+// See RpcWrapper for documentation.
+bool
+ClientTransactionTask::PrepareRpc::handleTransportError()
+{
+    // There was a transport-level failure. Flush cached state related
+    // to this session, and related to the object mappings.  The objects
+    // will all be retried when \c finish is called.
+    if (session.get() != NULL) {
+        ramcloud->clientContext->transportManager->flushSession(
+                session->getServiceLocator());
+        session = NULL;
+    }
+    retryRequest();
+    return true;
+}
+
+// See RpcWrapper for documentation.
+void
+ClientTransactionTask::PrepareRpc::send()
+{
+    reqHdr->ackId = ramcloud->rpcTracker.ackId();
+    state = IN_PROGRESS;
+    session->sendRequest(&request, response, this);
+}
+
+/**
+ * Append an operation to the end of this prepare rpc.
+ *
+ * \param opEntry
+ *      Handle to information about the operation to be appended.
+ */
+void
+ClientTransactionTask::PrepareRpc::appendOp(CommitCacheMap::iterator opEntry)
```

Comment:
I think the API might be a little cleaner if you passed in key and entry, rather than an iterator (this ties the method to a very particular approach being used by the caller).